### PR TITLE
Tablet throttler: starvation fix and consolidation of logic.

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

This PR addresses a starvation scenario in the tablet throttler. The scenario actually doesn't take place, due to an unintentional change. We fix the unintentional change & we fix the starvation scenario.

We also consolidate two different measurements of "last check" or "recent check":

1. `recentCheckValue`
2. `lastCheckTimeNano`

They are superfluous, because both serve to answer a similar question: "when was this throttler last checked?"

`lastCheckTimeNano` was updated in `check.go` at every check, and that's also where the aforementioned unintentional behavior takes place: this value should not be updated when the app is `"vitess"`. Right now in `main` this value does get updated when the app is `"vitess"`, which means the throttler's self-checks keep that value always recent. In turn, this prevents the throttler from going into _dormant mode_.

We introduce a new, single tracker for the checks, in the form of `recentCheckRateLimiter`. `RateLimiter` now not only produces ticks, and not only rate limits some action, but can also tell us the logical clock diff since last action invocation.

Using this, we can easily determine:

- Whether a check was made in the last 1-2 seconds
- Whether a check was made in the past `1min`

### Starvation prevention through `PRIMARY` stimulation

The fix to the starvation scenario is for the replica to proactively let the `PRIMARY` know that it has been _checked_. The `PRIMARY` still gets that information from the replicas when probing them. But now also, the replica invokes a `CheckThrottler()` directly on the `PRIMARY` with `"throttler-stimulator"` app name. This "stimulates" the `PRIMARY`, as it sees that it has just been _checked_, and therefore immediately renews heartbeat lease.

However, the replica does not overload the `PRIMARY` with these stimulations. Stimulations are rate limited to exactly _dormant period_ (`1min` at this time).

### Unit tests

We add unit tests to validate:

- Dormant mode is correct.
- `"vitess"` app checks does not affect "recently checked" state, neither or `PRIMARY` nor on replicas.
- `"throttler-stimulator"` affects "recently checked" state and causes heartbeat renewals.
- Replica checks invoke `CheckThrottler` on `PRIMARY` with `"throttler-stimulator"` app name.

## Related Issue(s)

https://github.com/vitessio/vitess/issues/15397

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vttablet/tabletserver/throttle/check.go [hunk 1]
```diff
@@ -148,20 +148,19 @@ func (check *ThrottlerCheck) Check(ctx context.Context, appName string, storeTyp
 	}
 
 	checkResult = check.checkAppMetricResult(ctx, appName, storeType, storeName, metricResultFunc, flags)
-	check.throttler.lastCheckTimeNano.Store(time.Now().UnixNano())
```
Comment (shlomi-noach) [inline go/vt/vttablet/tabletserver/throttle/check.go:151]: Removed from `check.,go`. Delegated to, and consolidated in `checkStore()` in `throttler.go`.

File: go/vt/vttablet/tabletserver/throttle/check.go [hunk 2]
```diff
@@ -148,20 +148,19 @@ func (check *ThrottlerCheck) Check(ctx context.Context, appName string, storeTyp
...
-		stats.GetOrNewCounter(fmt.Sprintf("ThrottlerCheckAny%s%sTotal", textutil.SingleWordCamel(storeType), textutil.SingleWordCamel(storeName)), "").Add(1)
-
-		if statusCode != http.StatusOK {
-			stats.GetOrNewCounter("ThrottlerCheckAnyError", "total number of failed checks").Add(1)
-			stats.GetOrNewCounter(fmt.Sprintf("ThrottlerCheckAny%s%sError", textutil.SingleWordCamel(storeType), textutil.SingleWordCamel(storeName)), "").Add(1)
-		}
-
-		check.throttler.markRecentApp(appName, remoteAddr)
-	}(checkResult.StatusCode)
+	if !throttlerapp.VitessName.Equals(appName) {
```
Comment (shlomi-noach) [inline go/vt/vttablet/tabletserver/throttle/check.go:151]: No metrics for `"vitess"` app.

**Comment by shlomi-noach** [comment]
Temprarily converting to `Draft` while running more checks.

File: go/vt/vttablet/tabletserver/throttle/throttler.go [hunk 3]
```diff
@@ -724,15 +755,42 @@ func (throttler *Throttler) Operate(ctx context.Context, wg *sync.WaitGroup) {
 			case <-mysqlCollectTicker.C:
 				if throttler.IsOpen() {
 					// frequent
+					// Always collect self metrics:
```
Comment (shlomi-noach) [inline go/vt/vttablet/tabletserver/throttle/throttler.go:758]: Now that this PR fixes the dormant issue, we also notice that any throttler should be collecting their own already-probed metrics. This comes at nearly no-cost. dormancy was made for the PRIMARY to not overwhelm network traffic when not needed, whilst collecting metrics from the replicas.

**Comment by shlomi-noach** [comment]
Upon adding https://github.com/vitessio/vitess/pull/15398/commits/b58bac6dabeba9fdc2b09a7924357c06dfdaf655, I realize this PR now solves a different starvation issue, perhaps even more dominant than the one it was meant to solve. Will open a new issue to describe it.

**Comment by shlomi-noach** [comment]
2nd starvation issue: https://github.com/vitessio/vitess/issues/15433

**Review by rohit-nayak-ps — APPROVED** [review]

**Review by dbussink — APPROVED** [review]

**Pull request merged**